src/util/any_ref.h

```diff
@@ -0,0 +1,82 @@
+/*******************************************************************\
+
+Module: Any reference - type-safe wrapper around void*
+
+Author: Diffblue ltd.
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_ANY_REF_H
+#define CPROVER_UTIL_ANY_REF_H
+
+#include <type_traits>
+#include <typeinfo>
+
+/// Any reference
+/// Stores a reference to any type
+/// Call .at<T>() to get the reference back
+class any_reft final
+{
+public:
+  template<typename T>
+  explicit any_reft(T &val):
+    m_type(&type_of<T>),
+    m_ptr(const_cast<typename std::remove_const<T>::type *>(&val)) { }
+
+  template<typename T>
+  any_reft &operator=(T &val)
+  {
+    m_type=&type_of<T>;
+    m_ptr=const_cast<typename std::remove_const<T>::type *>(&val);
+    return *this;
+  }
+
+  template<typename T>
+  bool operator==(const T &val) const
+  {
+    return (
+      m_type==&type_of<T>
+      || m_type==&type_of<typename std::remove_const<T>::type>)
+        && *static_cast<T *>(m_ptr)==val;
+  }
+
+  /// Convert to a reference of a specific type
+  /// Throw std::bad_cast if type doesn't match the type
+  template<typename T>
+  T &as()
```

Comment:
It seems that this can result in an unintended const_cast...

Reply:
Nope, there's even check for that in unit tests. Non-const can be cast to `const`, but not the other way around.